pkg/cmd/github-action-poller/main.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// Command github-action-poller monitors the status of provided GitHub
+// checks for a particular revision and reports their overall status.
+// It expects for all checks to have status "completed" and conclusion
+// "success" to exit zero.
+
+package main
+
+import (
+	"cmp"
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/google/go-github/v61/github"
+	"github.com/spf13/cobra"
+)
+
+var (
+	owner, repo, sha string
+	timeout          time.Duration
+	sleepDuration    time.Duration
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "github-action-poller",
+	Short: "Poll GitHub Action runs and wait for overall result",
+	Args:  cobra.MinimumNArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cmd.SilenceUsage = true
+
+		deadline := timeutil.Now().Add(timeout)
+
+		for {
+			shouldRetry, err := run(cmd.Context(), owner, repo, sha, args)
+			if err == nil {
+				return nil
+			}
+			if !shouldRetry {
+				return fmt.Errorf("failed: %w", err)
+			}
+			if timeutil.Now().After(deadline) {
+				return fmt.Errorf("timed out: %w", err)
+			}
+			log.Printf("sleeping for %s, will stop in %s", sleepDuration, timeutil.Until(deadline))
+			time.Sleep(sleepDuration)
+		}
+	},
+}
+
+func init() {
+	flags := rootCmd.Flags()
+	flags.StringVar(&owner, "owner", "", "repository owner")
+	flags.StringVar(&repo, "repo", "", "repository")
+	flags.StringVar(&sha, "sha", "", "SHA")
+	flags.DurationVar(&timeout, "timeout", 30*time.Minute, "overall program timeout")
+	flags.DurationVar(&sleepDuration, "sleep", 30*time.Second, "sleep between retries")
+
+	must := func(err error) {
+		if err != nil {
+			panic(err)
+		}
+	}
+	must(rootCmd.MarkFlagRequired("owner"))
+	must(rootCmd.MarkFlagRequired("repo"))
+	must(rootCmd.MarkFlagRequired("sha"))
+}
+
+func main() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func run(ctx context.Context, owner, repo, sha string, requiredChecks []string) (bool, error) {
+	statuses := make(map[string][]*github.CheckRun, len(requiredChecks))
+	for _, c := range requiredChecks {
+		statuses[c] = nil
+	}
+
+	client := github.NewClient(nil)
```

Comment:
This is unauthenticated, right? We'll hit rate limits even more quickly than usual. Maybe it doesn't make a difference if we do few enough requests.

Reply:
Correct. I can easily add auth if we want. Actually, it was there in the initial version. :)